examples/joystick/Joystick.cpp

```diff
@@ -0,0 +1,176 @@
+        return EXIT_FAILURE;
+
+    // Set up our string conversion parameters
+    sstr.precision(2);
+    sstr.setf(std::ios::fixed | std::ios::boolalpha);
+
+    // Set up our label-value sf::Text objects
+    texts["ID"].first.setPosition(5.f, 5.f);
+    texts["ID"].second.setPosition(80.f, 5.f);
+
+    for (unsigned int i = 0; i < sf::Joystick::AxisCount; ++i)
+    {
+        std::pair<sf::Text, sf::Text>& text = texts[axislabels[i]];
+
+        text.first.setPosition(5.f, 5.f + ((i + 2) * font.getLineSpacing(14)));
+        text.first.setString(std::string(axislabels[i]) + ":");
+
+        text.second.setPosition(80.f, 5.f + ((i + 2) * font.getLineSpacing(14)));
+    }
+
+    for (unsigned int i = 0; i < sf::Joystick::ButtonCount; ++i)
+    {
+        sstr.str("");
+        sstr << "Button " << i;
+        std::pair<sf::Text, sf::Text>& text = texts[sstr.str()];
+
+        text.first.setPosition(5.f, 5.f + ((sf::Joystick::AxisCount + i + 2) * font.getLineSpacing(14)));
+        text.first.setString(sstr.str() + ":");
+
+        text.second.setPosition(80.f, 5.f + ((sf::Joystick::AxisCount + i + 2) * font.getLineSpacing(14)));
+    }
+
+    for (Texts::iterator i = texts.begin(); i != texts.end(); ++i)
+    {
+        i->second.first.setFont(font);
+        i->second.first.setCharacterSize(14);
+        i->second.first.setFillColor(sf::Color::White);
+
+        i->second.second.setFont(font);
+        i->second.second.setCharacterSize(14);
+        i->second.second.setFillColor(sf::Color::White);
+    }
+
+    while (window.isOpen())
+    {
+        // Handle events
+        sf::Event event;
+        while (window.pollEvent(event))
+        {
+            // Window closed or escape key pressed: exit
+            if ((event.type == sf::Event::Closed) ||
+               ((event.type == sf::Event::KeyPressed) && (event.key.code == sf::Keyboard::Escape)))
+            {
+                window.close();
+                break;
+            }
+            else if ((event.type == sf::Event::JoystickButtonPressed) ||
+                     (event.type == sf::Event::JoystickButtonReleased) ||
+                     (event.type == sf::Event::JoystickMoved) ||
+                     (event.type == sf::Event::JoystickConnected))
+            {
+                // Update displayed joystick values
+                updateValues(event.joystickConnect.joystickId);
+            }
+            else if (event.type == sf::Event::JoystickDisconnected)
+            {
+                // Reset displayed joystick values to empty
+                for (Texts::iterator i = texts.begin(); i != texts.end(); ++i)
+                    i->second.second.setString("");
+
+                texts["ID"].first.setString("");
+            }
+        }
+
+        // Clear the window
+        window.clear();
+
+        // Draw the label-value sf::Text objects
+        for (Texts::const_iterator i = texts.begin(); i != texts.end(); ++i)
```

Comment:
If there's no controller connected it will still spit out basically placeholder text for the first controller. Maybe there should be some instruction like "No controllers could be found. Connect a gamepad or joystick to test it."?

Reply:
The text that is always output isn't placeholder text, they are simply the labels of the joystick items. When a joystick is connected, values will be displayed for the last joystick whose state was modified. When no joysticks are connected, no values are displayed.

Reply:
Yeah, but (at least to me) from a first look it's still not clear whether the program is stuck or just couldn't find any joystick. So maybe just rather than setting `ID` to nothing, set it to `<Not connected>`, if there's no controller? Similar not existing buttons/axes should possibly output "n/a" rather than nothing.